### ISSUE TITLE
NameError when getting cloudformation outputs

### ISSUE THREAD

**Issue by gdhbashton** [Jimdo/hiera-aws#26]
Hi,

On CentOS 7 after `gem install hiera-aws` I failed to execute a simple `info()` call from Puppet:

```
   info(hiera("cloudformation stack=root-NAT-10ZUOJCD35VC3 output=natsubnets"))
``` 

so I tried to make the same call using the hiera commandline tool directly:

```
[root@logstash-id9d54c3d manifests]# hiera -c /etc/puppet/hiera.yaml  "cloudformation stack=root-NAT-10ZUOJCD35VC3 output=natsubnets"
/usr/local/share/gems/gems/hiera-aws-0.6.0/lib/hiera/backend/aws_backend.rb:65:in `setup_aws_config': uninitialized constant Hiera::Backend::Aws_backend::AWS (NameError)
	from /usr/local/share/gems/gems/hiera-aws-0.6.0/lib/hiera/backend/aws_backend.rb:17:in `initialize'
	from /usr/share/ruby/vendor_ruby/hiera/backend.rb:205:in `new'
	from /usr/share/ruby/vendor_ruby/hiera/backend.rb:205:in `block in lookup'
	from /usr/share/ruby/vendor_ruby/hiera/backend.rb:203:in `each'
	from /usr/share/ruby/vendor_ruby/hiera/backend.rb:203:in `lookup'
	from /usr/share/ruby/vendor_ruby/hiera.rb:60:in `lookup'
	from /bin/hiera:225:in `<main>'
```

I don't understand the NameError because that stack and output name definitely exist:

```
[root@logstash-id9d54c3d ~]# aws --region eu-west-1 cloudformation describe-stacks --stack-name root-NAT-10ZUOJCD35VC3 --query Stacks[0].Outputs[0]
{
    "OutputKey": "natsubnets", 
    "OutputValue": "subnet-cc0eaabb,subnet-b7dd25ee,subnet-f3229d96"
}
```

My `/etc/puppet/hiera.yaml` is unremarkable:

```
---
:hierarchy:
  - "locations/%{location}/%{product}/%{env}/%{role}"
  - "locations/%{location}/%{product}/%{env}/common"
  - "locations/%{location}/%{product}/common"
  - "locations/%{location}/common"
  - "locations/common"
  - common
  - aws/elasticache
  - aws/rds
:backends:
  - yaml
  - aws
:yaml:
  :datadir: /etc/puppet/hiera
:aws:
  :region: eu-west-1
```

Any suggestions warmly welcomed!

**Comment by mlafeldt** [comment]
This looks like the aws-sdk gem is missing.

```
/usr/local/share/gems/gems/hiera-aws-0.6.0/lib/hiera/backend/aws_backend.rb:65:in `setup_aws_config': uninitialized constant Hiera::Backend::Aws_backend::AWS (NameError)
```

Can you check `gem list` to see a list of installed gems and their version?

**Comment by gdhbashton** [comment]
Ah, nice I actually have the output of `gem list --local` just *before* installing `hiera-aws`:

```
[root@logstash-id9d54c3d ~]# gem list --local

*** LOCAL GEMS ***

amq-protocol (1.9.2)
bigdecimal (1.2.0)
bunny (1.7.0)
carrot-top (0.0.7)
crack (0.4.2)
gpgme (2.0.8)
hiera (1.3.4)
hiera-gpg (1.1.0)
io-console (0.4.2)
json (1.7.7)
json_pure (1.8.2)
mime-types (2.4.3)
mini_portile (0.6.2)
mixlib-cli (1.5.0)
netrc (0.10.3)
psych (2.0.0)
rdoc (4.0.0)
rest-client (1.7.3)
safe_yaml (1.0.4)
sensu-plugin (1.1.0)
stomp (1.3.2)
timeout (0.0.1)
timeout-extensions (0.0.0)
```

And then the installation itself:

```
[root@logstash-id9d54c3d ~]# gem install hiera-aws
Fetching: multi_json-1.10.1.gem (100%)
Successfully installed multi_json-1.10.1
Fetching: multi_xml-0.5.5.gem (100%)
Successfully installed multi_xml-0.5.5
Fetching: builder-3.2.2.gem (100%)
Successfully installed builder-3.2.2
Fetching: jmespath-1.0.2.gem (100%)
Successfully installed jmespath-1.0.2
Fetching: aws-sdk-core-2.0.28.gem (100%)
Successfully installed aws-sdk-core-2.0.28
Fetching: aws-sdk-resources-2.0.28.gem (100%)
Successfully installed aws-sdk-resources-2.0.28
Fetching: aws-sdk-2.0.28.gem (100%)
Successfully installed aws-sdk-2.0.28
Fetching: hiera-aws-0.6.0.gem (100%)
Successfully installed hiera-aws-0.6.0
Parsing documentation for multi_json-1.10.1
Installing ri documentation for multi_json-1.10.1
Parsing documentation for multi_xml-0.5.5
Installing ri documentation for multi_xml-0.5.5
Parsing documentation for builder-3.2.2
Installing ri documentation for builder-3.2.2
Parsing documentation for jmespath-1.0.2
Installing ri documentation for jmespath-1.0.2
Parsing documentation for aws-sdk-core-2.0.28
Installing ri documentation for aws-sdk-core-2.0.28
Parsing documentation for aws-sdk-resources-2.0.28
Installing ri documentation for aws-sdk-resources-2.0.28
Parsing documentation for aws-sdk-2.0.28
Installing ri documentation for aws-sdk-2.0.28
Parsing documentation for hiera-aws-0.6.0
Installing ri documentation for hiera-aws-0.6.0
8 gems installed
```

Hm, `aws-sdk` defnitely got dragged in just fine:

```
[root@logstash-id9d54c3d ~]# gem list

*** LOCAL GEMS ***

amq-protocol (1.9.2)
aws-sdk (2.0.28)
aws-sdk-core (2.0.28)
aws-sdk-resources (2.0.28)
...
```

**Comment by mlafeldt** [comment]
I guess the problem is that we currently [do not pin the aws-sdk version to v1](https://github.com/Jimdo/hiera-aws/blob/5993cdec8331152d4f8cc4cf68072cc035a21643/hiera-aws.gemspec#L21). As a result, v2 gets installed, which we do not support yet.

Please give hiera-aws another try after installing v1 of the SDK:

```
gem install aws-sdk -v 1.63.0
```

**Comment by mlafeldt** [comment]
PR #27 is a potential fix.

**Comment by mlafeldt** [comment]
@gdhbashton Just released a bugfix version of hiera-aws (0.6.1), which resolves the aws-sdk issue.

```
$ docker run -it --rm ruby:2.2 /bin/bash
root@e4a98d2bed02:/# gem install hiera-aws
Fetching: mini_portile-0.6.2.gem (100%)
Successfully installed mini_portile-0.6.2
Fetching: nokogiri-1.6.6.2.gem (100%)
Building native extensions.  This could take a while...
Successfully installed nokogiri-1.6.6.2
Fetching: aws-sdk-v1-1.63.0.gem (100%)
Successfully installed aws-sdk-v1-1.63.0
Fetching: aws-sdk-1.63.0.gem (100%)
Successfully installed aws-sdk-1.63.0
Fetching: hiera-aws-0.6.1.gem (100%)
Successfully installed hiera-aws-0.6.1
5 gems installed
```

Thanks for reporting the bug!
